src/scheduler/connectivity.rs

```diff
@@ -0,0 +1,363 @@
+        if *lock == DetailedConnectivity::Connected || *lock == DetailedConnectivity::NotConfigured
+        {
+            *lock = DetailedConnectivity::InterruptingIdle;
+        }
+        drop(lock);
+
+        for state in &[&mvbox.state, &sentbox.state] {
+            let mut lock = state.connectivity.0.lock().await;
+            if *lock == DetailedConnectivity::Connected {
+                *lock = DetailedConnectivity::InterruptingIdle;
+            }
+        }
+    }
+    // No need to send ConnectivityChanged, the user-facing connectivity doesn't change because
+    // of what we do here.
+}
+
+impl fmt::Debug for ConnectivityStore {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(guard) = self.0.try_lock() {
+            write!(f, "ConnectivityStore {:?}", &*guard)
+        } else {
+            write!(f, "ConnectivityStore [LOCKED]")
+        }
+    }
+}
+
+impl Context {
+    /// Get the current connectivity, i.e. whether the device is connected to the IMAP server.
+    /// One of:
+    /// - DC_CONNECTIVITY_NOT_CONNECTED (1000-1999): Show e.g. the string "Not connected" or a red dot
+    /// - DC_CONNECTIVITY_CONNECTING (2000-2999): Show e.g. the string "Connecting…" or a yellow dot
+    /// - DC_CONNECTIVITY_WORKING (3000-3999): Show e.g. the string "Getting new messages" or a spinning wheel
+    /// - DC_CONNECTIVITY_CONNECTED (>=4000): Show e.g. the string "Connected" or a green dot
+    ///
+    /// We don't use exact values but ranges here so that we can split up
+    /// states into multiple states in the future.
+    ///
+    /// Meant as a rough overview that can be shown
+    /// e.g. in the title of the main screen.
+    ///
+    /// If the connectivity changes, a DC_EVENT_CONNECTIVITY_CHANGED will be emitted.
+    pub async fn get_connectivity(&self) -> Connectivity {
+        match &*self.scheduler.read().await {
+            Scheduler::Running {
+                inbox,
+                mvbox,
+                sentbox,
+                ..
+            } => {
+                let states = [&inbox.state, &mvbox.state, &sentbox.state];
+                let mut connectivities = Vec::new();
+                for s in &states {
+                    // TODO/QUESTION get_basic() locks a mutex, and above we called `scheduler.read()`. This means
+                    // that we will be holding two locks, which sounds like a great opportunity for
+                    // a deadlock.
+                    // Below (commented out, slightly outdated), I wrote another possible
+                    // version of this code which first clones all the ConnectivityStore's
+                    // (which are Arc's under the hood), then releases the scheduler-read-lock and only then
+                    // calls `get_basic()`. Would this be better? Or don't I have to worry about deadlocks here at all?
+                    // Same goes for get_connectivity_html().
```

Comment:
Are deadlocks a possible problem? I never worked a lot with mutexes before, so I thought I'd rather ask

Reply:
If some other "process" (async coroutine) does the opposite, and first calls `get_basic()` and then `scheduler.read()`, they can deadlock.

If possible, holding only one lock seems better, even at the cost of some cloning.